src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,133 @@
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1) == null ||
+            !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol method = invocation.TargetMethod;
+
+                DiagnosticDescriptor rule;
+                if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                }
+                else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                }
+                else
+                {
+                    return;
+                }
+
+                if (IsStreamMethod(method, stream) && HasUndesiredArguments(method))
+                {
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(rule));
+                }
+            }, OperationKind.Invocation);
+        }
+
+
+        private static bool IsStreamMethod(IMethodSymbol method, INamedTypeSymbol stream)
+        {
+            return method.ContainingType.Equals(stream) ||
+                (method.OverriddenMethod != null && IsStreamMethod(method.OverriddenMethod, stream));
```

Comment:
I don't expect any type hierarchies deep enough to blow out the stack here, but since this is tail-recursive it can easily be made iterative.  (Then jump ahead to the part where "Stream" isn't relevant to the method, and I think what you really want is)

```C#
private static bool IsDefinedBy(IMethodSymbol method, INamedTypeSymbol type)
{
    while (method.OverriddenMethod != null)
    {
        method = method.OverriddenMethod;
    }

    return method.ContainingType.Equals(type);
}
```